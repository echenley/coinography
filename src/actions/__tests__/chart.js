jest.mock('axios')

import axios from 'axios'
import {
  updateChartType,
  updateCurrency,
  updateGranularity,
  fetchChartData,
} from '../chart'

describe('chart actions', () => {
  describe('updateChartType', () => {
    it('returns correct action object', () => {
      expect(updateChartType('line')).toMatchSnapshot()
    })
  })

  describe('updateCurrency', () => {
    it('returns correct action object', () => {
      expect(updateCurrency('BTC-USD')).toMatchSnapshot()
    })
  })

  describe('updateGranularity', () => {
    it('returns correct action object', () => {
      expect(updateGranularity(300)).toMatchSnapshot()
    })
  })

  describe('fetchChartData', () => {
    it('dispatches correct actions on successful fetch', async () => {
      const response = { data: [] }
      axios.get.mockReturnValueOnce(Promise.resolve(response))

      const dispatch = jest.fn()
      const thunk = fetchChartData('BTC-USD', 300)

      const promise = thunk(dispatch)

      expect(dispatch).toHaveBeenCalledWith({
        type: 'FETCH_HISTORIC_DATA_LOADING',
        requestId: 1,
        silent: false,
      })

      expect(axios.get.mock.calls[0][0]).toMatchSnapshot(
        'calls correct api endpoint'
      )

      const returnValue = await promise

      expect(returnValue).toBe(response)
      expect(dispatch).toHaveBeenCalledWith({
        type: 'FETCH_HISTORIC_DATA_SUCCESS',
        data: response.data,
        requestId: 1,
      })
    })

    it('dispatches correct actions on failed fetch', async () => {
      const errorResponse = {}
      axios.get.mockReturnValueOnce(Promise.reject(errorResponse))

      const dispatch = jest.fn()
      const thunk = fetchChartData('BTC-USD', 300)

      const promise = thunk(dispatch)

      expect(dispatch).toHaveBeenCalledWith({
        type: 'FETCH_HISTORIC_DATA_LOADING',
        requestId: 2,
        silent: false,
      })

      let error
      try {
        await promise
      } catch (e) {
        error = e
      }

      expect(error).toBe(errorResponse)
      expect(dispatch).toHaveBeenCalledWith({
        type: 'FETCH_HISTORIC_DATA_FAILURE',
        error: errorResponse,
        requestId: 2,
      })
    })
  })
})
