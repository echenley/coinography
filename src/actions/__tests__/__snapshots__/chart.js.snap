// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chart actions fetchChartData dispatches correct actions on successful fetch: calls correct api endpoint 1`] = `"https://api.gdax.com/products/BTC-USD/candles?granularity=300"`;

exports[`chart actions updateChartType returns correct action object 1`] = `
Object {
  "chartType": "line",
  "type": "UPDATE_CHART_TYPE",
}
`;

exports[`chart actions updateCurrency returns correct action object 1`] = `
Object {
  "currency": "BTC-USD",
  "type": "UPDATE_CURRENCY",
}
`;

exports[`chart actions updateGranularity returns correct action object 1`] = `
Object {
  "granularity": 300,
  "type": "UPDATE_GRANULARITY",
}
`;
